File: plugins/nkn-sdk-flutter/ios/Classes/impl/Wallet.swift
import Nkn

class Wallet : ChannelBase, IChannelHandler, FlutterStreamHandler {

    let CHANNEL_NAME = "org.nkn.sdk/wallet"
    var methodChannel: FlutterMethodChannel?
    var eventSink: FlutterEventSink?

    let walletQueue = DispatchQueue(label: "org.nkn.sdk/wallet/queue", qos: .default, attributes: .concurrent)
    private var walletWorkItem: DispatchWorkItem?

    let walletMoneyQueue = DispatchQueue(label: "org.nkn.sdk/wallet/money/queue", qos: .default, attributes: .concurrent)
    private var walletMoneyWorkItem: DispatchWorkItem?

    let walletEventQueue = DispatchQueue(label: "org.nkn.sdk/wallet/event/queue", qos: .default, attributes: .concurrent)
    private var walletEventWorkItem: DispatchWorkItem?

    func install(binaryMessenger: FlutterBinaryMessenger) {
        self.methodChannel = FlutterMethodChannel(name: CHANNEL_NAME, binaryMessenger: binaryMessenger)
        self.methodChannel?.setMethodCallHandler(handle)
    }

    func uninstall() {
        self.methodChannel?.setMethodCallHandler(nil)
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        // eventSink = nil
        return nil
    }

    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method{
        case "measureSeedRPCServer":
            measureSeedRPCServer(call, result: result)
        case "create":
            create(call, result: result)
        case "restore" :
            restore(call, result: result)
        case "pubKeyToWalletAddr":
            pubKeyToWalletAddr(call, result: result)
        case "getBalance" :
            getBalance(call, result: result)
        case "transfer":
            transfer(call, result: result)
        case "subscribe":
            subscribe(call, result: result)
        case "unsubscribe":
            unsubscribe(call, result: result)
        case "getSubscribersCount":
            getSubscribersCount(call, result: result)
        case "getSubscribers":
            getSubscribers(call, result: result)
        case "getSubscription":
            getSubscription(call, result: result)
        case "getHeight":
            getHeight(call, result: result)
        case "getNonce":
            getNonce(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func measureSeedRPCServer(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let seedRpc = args["seedRpc"] as? [String]
        let timeout = args["timeout"] as? Int32 ?? 3000

        walletWorkItem = DispatchWorkItem {
            var seedRPCServerAddr = NkngomobileNewStringArrayFromString(nil)
            for (_, v) in seedRpc!.enumerated() {
                seedRPCServerAddr?.append(v)
            }
            seedRPCServerAddr = NknMeasureSeedRPCServer(seedRPCServerAddr as! NkngomobileStringArray, timeout, nil)

            var seedRPCServerAddrs = [String]()
            let elements = seedRPCServerAddr?.join(",").split(separator: ",")
            if elements != nil && !(elements!.isEmpty) {
                for element in elements! {
                    if !(element.isEmpty) {
                        seedRPCServerAddrs.append("\(element)")
                    }
                }
            }

            var resp:[String:Any] = [String:Any]()
            resp["seedRPCServerAddrList"] = seedRPCServerAddrs
            self.resultSuccess(result: result, resp: resp)
        }
        walletQueue.async(execute: walletWorkItem!)
    }

    private func create(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let seed = args["seed"] as? FlutterStandardTypedData
        let password = args["password"] as? String ?? ""
        let seedRpc = args["seedRpc"] as? [String]

        let config = NknWalletConfig()
        config.password = password
        if(seedRpc != nil) {
            config.seedRPCServerAddr = NkngomobileStringArray(from: nil)
            for (_, v) in seedRpc!.enumerated() {
                config.seedRPCServerAddr?.append(v)
            }
        }
        // config.rpcConcurrency = 4

        walletWorkItem = DispatchWorkItem {
            var error: NSError?
            let account:NknAccount? = NknNewAccount(seed?.data, &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }
            let wallet = NknWallet(account, config: config)

            var resp:[String:Any] = [String:Any]()
            resp["address"] = wallet?.address()
            resp["keystore"] = wallet?.toJSON(nil)
            resp["publicKey"] = wallet?.pubKey()
            resp["seed"] = wallet?.seed()
            self.resultSuccess(result: result, resp: resp)
        }
        walletQueue.async(execute: walletWorkItem!)
    }

    private func restore(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let keystore = args["keystore"] as? String
        let password = args["password"] as? String ?? ""
        let seedRpc = args["seedRpc"] as? [String]

        if(keystore == nil) {
            result(nil)
            return
        }

        let config = NknWalletConfig()
        config.password = password
        if(seedRpc != nil) {
            config.seedRPCServerAddr = NkngomobileNewStringArrayFromString(nil)
            for (_, v) in seedRpc!.enumerated() {
                config.seedRPCServerAddr?.append(v)
            }
        }
        // config.rpcConcurrency = 4

        walletWorkItem = DispatchWorkItem {
            var error: NSError?
            let wallet = NknWalletFromJSON(keystore, config, &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }

            var resp:[String:Any] = [String:Any]()
            resp["address"] = wallet?.address()
            resp["keystore"] = wallet?.toJSON(nil)
            resp["publicKey"] = wallet?.pubKey()
            resp["seed"] = wallet?.seed()
            self.resultSuccess(result: result, resp: resp)
        }
        walletQueue.async(execute: walletWorkItem!)
    }

    private func pubKeyToWalletAddr(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let publicKey = args["publicKey"] as! String

        walletWorkItem = DispatchWorkItem {
            var error: NSError?
            let address = NknPubKeyToWalletAddr(Data(hex: publicKey), &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }
            self.resultSuccess(result: result, resp: address)
        }
        walletQueue.async(execute: walletWorkItem!)
    }

    func getBalance(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let address = args["address"] as? String
        let seedRpc = args["seedRpc"] as? [String]

        let config = NknWalletConfig()
        if(seedRpc != nil) {
            config.seedRPCServerAddr = NkngomobileNewStringArrayFromString(nil)
            for (_, v) in seedRpc!.enumerated() {
                config.seedRPCServerAddr?.append(v)
            }
        }
        // else {
        //     config.seedRPCServerAddr = NknStringArray.init(from: "https://mainnet-rpc-node-0001.nkn.org/mainnet/api/wallet")
        // }
        // config.rpcConcurrency = 4

        walletMoneyWorkItem = DispatchWorkItem {
            var error: NSError?
            let account = NknAccount(NknRandomBytes(32, &error))
            if(error != nil) {
                self.resultError(result: result, error: error)
                return
            }

            let wallet = NknWallet(account, config: config)
            do {
                let balance: NknAmount? = try wallet?.balance(byAddress: address)

                self.resultSuccess(result: result, resp: Double(balance!.string()))
                return
            } catch let error {
                self.resultError(result: result, error: error)
                return
            }
        }
        walletMoneyQueue.async(execute: walletMoneyWorkItem!)
    }

    func transfer(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let seed = args["seed"] as? FlutterStandardTypedData
        let address = args["address"] as? String
        let amount = args["amount"] as? String
        let fee = args["fee"] as! String
        let nonce = args["nonce"] as? Int
        let attributes = args["attributes"] as? FlutterStandardTypedData
        let seedRpc = args["seedRpc"] as? [String]

        let config = NknWalletConfig()
        if(seedRpc != nil) {
            config.seedRPCServerAddr = NkngomobileNewStringArrayFromString(nil)
            for (_, v) in seedRpc!.enumerated() {
                config.seedRPCServerAddr?.append(v)
            }
        }
        // else {
        //     config.seedRPCServerAddr = NknStringArray.init(from: "https://mainnet-rpc-node-0001.nkn.org/mainnet/api/wallet")
        // }
        // config.rpcConcurrency = 4

        walletMoneyWorkItem = DispatchWorkItem {
            var error: NSError?
            let account:NknAccount? = NknNewAccount(seed?.data, &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }

            let wallet = NknNewWallet(account, config, &error)
            if (error != nil) {
                self.resultError(result: result,error: error)
                return
            }

            let transactionConfig: NknTransactionConfig = NknTransactionConfig()
            transactionConfig.fee = fee
            if (nonce != nil) {
                transactionConfig.nonce = Int64(nonce!)
                transactionConfig.fixNonce = true
            }
            if(attributes != nil){
                transactionConfig.attributes = attributes?.data
            }

            let hash = wallet?.transfer(address, amount: amount, config: transactionConfig, error: &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }

            self.resultSuccess(result: result, resp: hash)
            return
        }
        walletMoneyQueue.async(execute: walletMoneyWorkItem!)
    }

    func subscribe(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let seed = args["seed"] as? FlutterStandardTypedData
        let identifier = args["identifier"] as? String ?? ""
        let topic = args["topic"] as! String
        let duration = args["duration"] as! Int
        let meta = args["meta"] as? String
        let fee = args["fee"] as? String ?? "0"
        let nonce = args["nonce"] as? Int
        let seedRpc = args["seedRpc"] as? [String]

        let config = NknWalletConfig()
        if(seedRpc != nil) {
            config.seedRPCServerAddr = NkngomobileNewStringArrayFromString(nil)
            for (_, v) in seedRpc!.enumerated() {
                config.seedRPCServerAddr?.append(v)
            }
        }
        // config.rpcConcurrency = 4

        walletMoneyWorkItem = DispatchWorkItem {
            var error: NSError?
            let account:NknAccount? = NknNewAccount(seed?.data, &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }

            let wallet = NknNewWallet(account, config, &error)
            if (error != nil) {
                self.resultError(result: result,error: error)
                return
            }

            let transactionConfig: NknTransactionConfig = NknTransactionConfig()
            transactionConfig.fee = fee
            if (nonce != nil) {
                transactionConfig.nonce = Int64(nonce!)
                transactionConfig.fixNonce = true
            }


            let hash = wallet?.subscribe(identifier, topic: topic, duration: duration, meta: meta, config: transactionConfig, error: &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }

            self.resultSuccess(result: result, resp: hash)
            return
        }
        walletMoneyQueue.async(execute: walletMoneyWorkItem!)
    }

    func unsubscribe(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let seed = args["seed"] as? FlutterStandardTypedData
        let identifier = args["identifier"] as? String ?? ""
        let topic = args["topic"] as! String
        let fee = args["fee"] as? String ?? "0"
        let nonce = args["nonce"] as? Int
        let seedRpc = args["seedRpc"] as? [String]

        let config = NknWalletConfig()
        if(seedRpc != nil) {
            config.seedRPCServerAddr = NkngomobileNewStringArrayFromString(nil)
            for (_, v) in seedRpc!.enumerated() {
                config.seedRPCServerAddr?.append(v)
            }
        }
        // config.rpcConcurrency = 4

        walletMoneyWorkItem = DispatchWorkItem {
            var error: NSError?
            let account:NknAccount? = NknNewAccount(seed?.data, &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }

            let wallet = NknNewWallet(account, config, &error)
            if (error != nil) {
                self.resultError(result: result,error: error)
                return
            }

            let transactionConfig: NknTransactionConfig = NknTransactionConfig()
            transactionConfig.fee = fee
            if (nonce != nil) {
                transactionConfig.nonce = Int64(nonce!)
                transactionConfig.fixNonce = true
            }


            let hash = wallet?.unsubscribe(identifier, topic: topic, config: transactionConfig, error: &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }

            self.resultSuccess(result: result, resp: hash)
            return
        }
        walletMoneyQueue.async(execute: walletMoneyWorkItem!)
    }

    private func getSubscribers(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let topic = args["topic"] as! String
        let offset = args["offset"] as? Int ?? 0
        let limit = args["limit"] as? Int ?? 0
        let meta = args["meta"] as? Bool ?? true
        let txPool = args["txPool"] as? Bool ?? true
        let subscriberHashPrefix = args["subscriberHashPrefix"] as? FlutterStandardTypedData
        let seedRpc = args["seedRpc"] as? [String]

        let config = NknWalletConfig()
        if(seedRpc != nil) {
            config.seedRPCServerAddr = NkngomobileNewStringArrayFromString(nil)
            for (_, v) in seedRpc!.enumerated() {
                config.seedRPCServerAddr?.append(v)
            }
        }
        // config.rpcConcurrency = 4

        walletEventWorkItem = DispatchWorkItem {
            var error: NSError?
            let res: NknSubscribers? = NknGetSubscribers(topic, offset, limit, meta, txPool, subscriberHashPrefix?.data, config, &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }

            let mapPro = MapProtocol()
            res?.subscribers?.range(mapPro)

            self.resultSuccess(result: result, resp: mapPro.result)
            return
        }
        walletEventQueue.async(execute: walletEventWorkItem!)
    }

    private func getSubscribersCount(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let topic = args["topic"] as! String
        let subscriberHashPrefix = args["subscriberHashPrefix"] as? FlutterStandardTypedData
        let seedRpc = args["seedRpc"] as? [String]

        let config = NknWalletConfig()
        if(seedRpc != nil) {
            config.seedRPCServerAddr = NkngomobileNewStringArrayFromString(nil)
            for (_, v) in seedRpc!.enumerated() {
                config.seedRPCServerAddr?.append(v)
            }
        }
        // config.rpcConcurrency = 4

        walletEventWorkItem = DispatchWorkItem {
            var count: Int = 0
            var error: NSError?
            NknGetSubscribersCount(topic, subscriberHashPrefix?.data, config, &count, &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }

            self.resultSuccess(result: result, resp: count)
            return
        }
        walletEventQueue.async(execute: walletEventWorkItem!)
    }

    private func getSubscription(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        // let _id = args["_id"] as! String
        let topic = args["topic"] as! String
        let subscriber = args["subscriber"] as! String
        let seedRpc = args["seedRpc"] as? [String]

        let config = NknWalletConfig()
        if(seedRpc != nil) {
            config.seedRPCServerAddr = NkngomobileNewStringArrayFromString(nil)
            for (_, v) in seedRpc!.enumerated() {
                config.seedRPCServerAddr?.append(v)
            }
        }
        // config.rpcConcurrency = 4

        walletEventWorkItem = DispatchWorkItem {
            var error: NSError?
            let res: NknSubscription? = NknGetSubscription(topic, subscriber, config, &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }

            var resp: [String: Any] = [String: Any]()
            resp["meta"] = res?.meta
            resp["expiresAt"] = res?.expiresAt
            self.resultSuccess(result: result, resp: resp)
            return
        }
        walletEventQueue.async(execute: walletEventWorkItem!)
    }

    private func getHeight(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let seedRpc = args["seedRpc"] as? [String]

        let config = NknWalletConfig()
        if(seedRpc != nil) {
            config.seedRPCServerAddr = NkngomobileNewStringArrayFromString(nil)
            for (_, v) in seedRpc!.enumerated() {
                config.seedRPCServerAddr?.append(v)
            }
        }
        // config.rpcConcurrency = 4

        walletEventWorkItem = DispatchWorkItem {
            var height: Int32 = 0
            var error: NSError?
            NknGetHeight(config, &height, &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }

            self.resultSuccess(result: result, resp: height)
            return
        }
        walletEventQueue.async(execute: walletEventWorkItem!)
    }

    private func getNonce(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: Any]
        let address = args["address"] as! String
        let txPool = args["txPool"] as? Bool ?? true
        let seedRpc = args["seedRpc"] as? [String]

        let config = NknWalletConfig()
        if(seedRpc != nil) {
            config.seedRPCServerAddr = NkngomobileNewStringArrayFromString(nil)
            for (_, v) in seedRpc!.enumerated() {
                config.seedRPCServerAddr?.append(v)
            }
        }
        // config.rpcConcurrency = 4

        walletEventWorkItem = DispatchWorkItem {
            var nonce: Int64 = 0
            var error: NSError?
            NknGetNonce(address, txPool, config, &nonce, &error)
            if (error != nil) {
                self.resultError(result: result, error: error)
                return
            }

            self.resultSuccess(result: result, resp: nonce)
            return
        }
        walletEventQueue.async(execute: walletEventWorkItem!)
    }
}

File: plugins/nkn-sdk-flutter/android/src/main/kotlin/org/nkn/sdk/impl/Wallet.kt
package org.nkn.sdk.impl

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import nkn.Nkn
import nkn.RPCConfig
import nkn.TransactionConfig
import nkn.WalletConfig
import nkngolib.Nkngolib
import nkngomobile.StringArray
import org.bouncycastle.util.encoders.Hex
import org.nkn.sdk.IChannelHandler

class Wallet : IChannelHandler, MethodChannel.MethodCallHandler, EventChannel.StreamHandler, ViewModel() {
    companion object {
        val CHANNEL_NAME = "org.nkn.sdk/wallet"
    }

    lateinit var methodChannel: MethodChannel
    var eventSink: EventChannel.EventSink? = null

    override fun install(binaryMessenger: BinaryMessenger) {
        methodChannel = MethodChannel(binaryMessenger, CHANNEL_NAME)
        methodChannel.setMethodCallHandler(this)
    }

    override fun uninstall() {
        methodChannel.setMethodCallHandler(null)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "measureSeedRPCServer" -> {
                measureSeedRPCServer(call, result)
            }
            "create" -> {
                create(call, result)
            }
            "restore" -> {
                restore(call, result)
            }
            "pubKeyToWalletAddr" -> {
                pubKeyToWalletAddr(call, result)
            }
            "getBalance" -> {
                getBalance(call, result)
            }
            "transfer" -> {
                transfer(call, result)
            }
            "subscribe" -> {
                subscribe(call, result)
            }
            "unsubscribe" -> {
                unsubscribe(call, result)
            }
            "getSubscribersCount" -> {
                getSubscribersCount(call, result)
            }
            "getSubscribers" -> {
                getSubscribers(call, result)
            }
            "getSubscription" -> {
                getSubscription(call, result)
            }
            "getHeight" -> {
                getHeight(call, result)
            }
            "getNonce" -> {
                getNonce(call, result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun measureSeedRPCServer(call: MethodCall, result: MethodChannel.Result) {
        val seedRpc = call.argument<ArrayList<String>?>("seedRpc") ?: arrayListOf()
        val timeout = call.argument<Int>("timeout") ?: 3000

        var seedRPCServerAddr = StringArray(null)
        for (addr in seedRpc) {
            seedRPCServerAddr.append(addr)
        }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                seedRPCServerAddr = Nkngolib.measureSeedRPCServer(seedRPCServerAddr, timeout)

                val seedRPCServerAddrs = arrayListOf<String>()
                val elements = seedRPCServerAddr.join(",").split(",")
                for (element in elements) {
                    if (element.isNotEmpty()) {
                        seedRPCServerAddrs.add(element)
                    }
                }

                val resp = hashMapOf(
                    "seedRPCServerAddrList" to seedRPCServerAddrs
                )
                resultSuccess(result, resp)
                return@launch
            } catch (e: Throwable) {
                resultError(result, e)
                return@launch
            }
        }
    }

    private fun create(call: MethodCall, result: MethodChannel.Result) {
        val seed = call.argument<ByteArray>("seed") ?: Nkn.randomBytes(32)
        val password = call.argument<String>("password") ?: ""
        val seedRpc = call.argument<ArrayList<String>?>("seedRpc")

        val config = WalletConfig()
        config.password = password
        if (seedRpc != null) {
            config.seedRPCServerAddr = StringArray(null)
            for (addr in seedRpc) {
                config.seedRPCServerAddr.append(addr)
            }
        }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val account = Nkn.newAccount(seed)
                val wallet = Nkn.newWallet(account, config)

                val resp = hashMapOf(
                    "address" to wallet.address(),
                    "keystore" to wallet.toJSON(),
                    "publicKey" to wallet.pubKey(),
                    "seed" to wallet.seed()
                )
                resultSuccess(result, resp)
                return@launch
            } catch (e: Throwable) {
                resultError(result, e)
                return@launch
            }
        }
    }

    private fun restore(call: MethodCall, result: MethodChannel.Result) {
        val keystore = call.argument<String>("keystore")
        val password = call.argument<String>("password") ?: ""
        val seedRpc = call.argument<ArrayList<String>?>("seedRpc")

        if (keystore == null) {
            result.success(null)
            return
        }

        val config = WalletConfig()
        config.password = password
        if (seedRpc != null) {
            config.seedRPCServerAddr = StringArray(null)
            for (addr in seedRpc) {
                config.seedRPCServerAddr.append(addr)
            }
        }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val wallet = Nkn.walletFromJSON(keystore, config)

                val resp = hashMapOf(
                    "address" to wallet.address(),
                    "keystore" to wallet?.toJSON(),
                    "publicKey" to wallet.pubKey(),
                    "seed" to wallet.seed()
                )
                resultSuccess(result, resp)
                return@launch
            } catch (e: Throwable) {
                resultError(result, e)
                return@launch
            }
        }
    }

    private fun pubKeyToWalletAddr(call: MethodCall, result: MethodChannel.Result) {
        val pubkey = call.argument<String>("publicKey")

        viewModelScope.launch(Dispatchers.IO) {
            val addr = Nkn.pubKeyToWalletAddr(Hex.decode(pubkey))
            resultSuccess(result, addr)
            return@launch
        }
    }

    private fun getBalance(call: MethodCall, result: MethodChannel.Result) {
        val address = call.argument<String>("address")
        val seedRpc = call.argument<ArrayList<String>?>("seedRpc")
        val account = Nkn.newAccount(Nkn.randomBytes(32))

        val config = WalletConfig()
        if (seedRpc != null) {
            config.seedRPCServerAddr = StringArray(null)
            for (addr in seedRpc) {
                config.seedRPCServerAddr.append(addr)
            }
        }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val wallet = Nkn.newWallet(account, config)
                val balance = wallet.balanceByAddress(address).toString()

                resultSuccess(result, balance.toDouble())
                return@launch
            } catch (e: Throwable) {
                resultError(result, e)
                return@launch
            }
        }
    }

    private fun transfer(call: MethodCall, result: MethodChannel.Result) {
        val seed = call.argument<ByteArray>("seed")
        val address = call.argument<String>("address")
        val amount = call.argument<String>("amount") ?: "0"
        val fee = call.argument<String>("fee") ?: "0"
        val nonce = call.argument<Int>("nonce")
        val attributes = call.argument<ByteArray>("attributes")
        val seedRpc = call.argument<ArrayList<String>?>("seedRpc")

        val config = WalletConfig()
        if (seedRpc != null) {
            config.seedRPCServerAddr = StringArray(null)
            for (addr in seedRpc) {
                config.seedRPCServerAddr.append(addr)
            }
        }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val account = Nkn.newAccount(seed)
                val wallet = Nkn.newWallet(account, config)
                val transactionConfig = TransactionConfig()
                transactionConfig.fee = fee
                if (nonce != null) {
                    transactionConfig.nonce = nonce.toLong()
                    transactionConfig.fixNonce = true
                }
                if (attributes != null) {
                    transactionConfig.attributes = attributes
                }
                val hash = wallet.transfer(address, amount, transactionConfig)

                resultSuccess(result, hash)
                return@launch
            } catch (e: Throwable) {
                resultError(result, e)
                return@launch
            }
        }
    }

    private fun subscribe(call: MethodCall, result: MethodChannel.Result) {
        val seed = call.argument<ByteArray>("seed")
        val identifier = call.argument<String>("identifier") ?: ""
        val topic = call.argument<String>("topic")!!
        val duration = call.argument<Int>("duration")!!
        val meta = call.argument<String>("meta")
        val fee = call.argument<String>("fee") ?: "0"
        val nonce = call.argument<Int>("nonce")
        val seedRpc = call.argument<ArrayList<String>?>("seedRpc")

        val transactionConfig = TransactionConfig()
        transactionConfig.fee = fee
        if (nonce != null) {
            transactionConfig.nonce = nonce.toLong()
            transactionConfig.fixNonce = true
        }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val config = WalletConfig()
                if (seedRpc != null) {
                    config.seedRPCServerAddr = StringArray(null)
                    for (addr in seedRpc) {
                        config.seedRPCServerAddr.append(addr)
                    }
                }
                val account = Nkn.newAccount(seed)
                val wallet = Nkn.newWallet(account, config)
                val hash = wallet.subscribe(identifier, topic, duration.toLong(), meta, transactionConfig)

                resultSuccess(result, hash)
                return@launch
            } catch (e: Throwable) {
                resultError(result, e)
                return@launch
            }
        }
    }

    private fun unsubscribe(call: MethodCall, result: MethodChannel.Result) {
        val seed = call.argument<ByteArray>("seed")
        val identifier = call.argument<String>("identifier") ?: ""
        val topic = call.argument<String>("topic")!!
        val fee = call.argument<String>("fee") ?: "0"
        val nonce = call.argument<Int>("nonce")
        val seedRpc = call.argument<ArrayList<String>?>("seedRpc")

        val transactionConfig = TransactionConfig()
        transactionConfig.fee = fee
        if (nonce != null) {
            transactionConfig.nonce = nonce.toLong()
            transactionConfig.fixNonce = true
        }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val config = WalletConfig()
                if (seedRpc != null) {
                    config.seedRPCServerAddr = StringArray(null)
                    for (addr in seedRpc) {
                        config.seedRPCServerAddr.append(addr)
                    }
                }
                val account = Nkn.newAccount(seed)
                val wallet = Nkn.newWallet(account, config)
                val hash = wallet.unsubscribe(identifier, topic, transactionConfig)

                resultSuccess(result, hash)
                return@launch
            } catch (e: Exception) {
                resultError(result, e)
                return@launch
            }
        }
    }

    private fun getSubscribers(call: MethodCall, result: MethodChannel.Result) {
        val topic = call.argument<String>("topic")!!
        val offset = call.argument<Int>("offset") ?: 0
        val limit = call.argument<Int>("limit") ?: 0
        val meta = call.argument<Boolean>("meta") ?: true
        val txPool = call.argument<Boolean>("txPool") ?: true
        val seedRpc = call.argument<ArrayList<String>?>("seedRpc")
        val subscriberHashPrefix = call.argument<ByteArray>("subscriberHashPrefix")

        val config = RPCConfig()
        if (seedRpc != null) {
            config.seedRPCServerAddr = StringArray(null)
            for (addr in seedRpc) {
                config.seedRPCServerAddr.append(addr)
            }
        }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val subscribers = Nkn.getSubscribers(topic, offset.toLong(), limit.toLong(), meta, txPool, subscriberHashPrefix, config)
                val resp = hashMapOf<String, String>()
                subscribers.subscribers.range { addr, value ->
                    resp[addr] = value?.trim() ?: ""
                    true
                }

                resultSuccess(result, resp)
                return@launch
            } catch (e: Exception) {
                resultError(result, e)
                return@launch
            }
        }
    }

    private fun getSubscription(call: MethodCall, result: MethodChannel.Result) {
        val topic = call.argument<String>("topic")!!
        val subscriber = call.argument<String>("subscriber")!!
        val seedRpc = call.argument<ArrayList<String>?>("seedRpc")

        val config = RPCConfig()
        if (seedRpc != null) {
            config.seedRPCServerAddr = StringArray(null)
            for (addr in seedRpc) {
                config.seedRPCServerAddr.append(addr)
            }
        }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val subscription = Nkn.getSubscription(topic, subscriber, config)
                val resp = hashMapOf(
                    "meta" to subscription.meta,
                    "expiresAt" to subscription.expiresAt
                )

                resultSuccess(result, resp)
                return@launch
            } catch (e: Exception) {
                resultError(result, e)
                return@launch
            }
        }
    }

    private fun getSubscribersCount(call: MethodCall, result: MethodChannel.Result) {
        val topic = call.argument<String>("topic")!!
        val seedRpc = call.argument<ArrayList<String>?>("seedRpc")
        val subscriberHashPrefix = call.argument<ByteArray>("subscriberHashPrefix")

        val config = RPCConfig()
        if (seedRpc != null) {
            config.seedRPCServerAddr = StringArray(null)
            for (addr in seedRpc) {
                config.seedRPCServerAddr.append(addr)
            }
        }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val count = Nkn.getSubscribersCount(topic, subscriberHashPrefix, config)

                resultSuccess(result, count)
                return@launch
            } catch (e: Exception) {
                resultError(result, e)
                return@launch
            }
        }
    }

    private fun getHeight(call: MethodCall, result: MethodChannel.Result) {
        val seedRpc = call.argument<ArrayList<String>?>("seedRpc")

        val config = RPCConfig()
        if (seedRpc != null) {
            config.seedRPCServerAddr = StringArray(null)
            for (addr in seedRpc) {
                config.seedRPCServerAddr.append(addr)
            }
        }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val height = Nkn.getHeight(config)

                resultSuccess(result, height)
                return@launch
            } catch (e: Exception) {
                resultError(result, e)
                return@launch
            }
        }
    }

    private fun getNonce(call: MethodCall, result: MethodChannel.Result) {
        val address = call.argument<String>("address")
        val txPool = call.argument<Boolean>("txPool") ?: true
        val seedRpc = call.argument<ArrayList<String>?>("seedRpc")

        val config = RPCConfig()
        if (seedRpc != null) {
            config.seedRPCServerAddr = StringArray(null)
            for (addr in seedRpc) {
                config.seedRPCServerAddr.append(addr)
            }
        }

        viewModelScope.launch(Dispatchers.IO) {
            try {
                val nonce = Nkn.getNonce(address, txPool, config)

                resultSuccess(result, nonce)
                return@launch
            } catch (e: Exception) {
                resultError(result, e)
                return@launch
            }
        }
    }
}